### ISSUE TITLE
support domains

### ISSUE THREAD

**Issue by mundodisco8** [rkitover/windows-alt-sshd-msys2#2]
Hi, 

I just run your script, but I can't make it work. Terminal won't connect

![image](https://user-images.githubusercontent.com/8402707/85332188-60979c00-b4cf-11ea-80e7-40182f61b9fd.png)

Sadly I closed the MSYS window where I ran the script so I don't have the output, but if I run it again, I get this:

![image](https://user-images.githubusercontent.com/8402707/85332262-8a50c300-b4cf-11ea-8b1f-0b40b5d71eb5.png)

Any ideas?

**Comment by rkitover** [comment]
Are you joined to a domain? I have not tested with domains.

**Comment by mundodisco8** [comment]
Ah! I might! this is my work computer...

**Comment by rkitover** [comment]
I will try to look at this, but it may non-trivial for me to set up a testing environment at home with a domain controller.

I'd appreciate any help with this for anyone who knows any shell scripting.

**Comment by rkitover** [comment]
@mundodisco8 can you try latest master please, I made a minor change, I don't know if it will fix all the issues with domains but it's worth a try.

**Comment by mundodisco8** [comment]
Hi,

Thanks for helping me with this one. I tried the latest and sadly it doesn't work. This is the output

```text
# bash msys2-alt-sshd-setup.sh
:: Synchronising package databases...
 mingw32                                     467.0 KiB   491 KiB/s 00:01 [#######################################] 100%
 mingw32.sig                                 119.0   B  0.00   B/s 00:00 [#######################################] 100%
 mingw64                                     469.7 KiB  1441 KiB/s 00:00 [#######################################] 100%
 mingw64.sig                                 119.0   B  0.00   B/s 00:00 [#######################################] 100%
 msys                                        187.7 KiB  2040 KiB/s 00:00 [#######################################] 100%
 msys.sig                                    119.0   B  0.00   B/s 00:00 [#######################################] 100%
warning: openssh-8.3p1-1 is up to date -- skipping
warning: cygrunsrv-1.62-2 is up to date -- skipping
warning: mingw-w64-x86_64-editrights-1.03-3 is up to date -- skipping
 there is nothing to do
The password does not meet the password policy requirements. Check the minimum password length, password complexity and password history requirements.

More help is available by typing NET HELPMSG 2245.

ERROR: Unable to create Windows user sshd_server_msys2
```

I don't want to take too much of your time, so if you don't feel like fixing it, don't worry, I totally understand.

Cheers

**Comment by rkitover** [comment]
Looks like an issue with the generated password not meeting the domain password requirements, I'll change that.

**Comment by rkitover** [comment]
@mundodisco8 can you please try master again?

**Comment by mundodisco8** [comment]
I think my computer is very stubborn

```text
# bash msys2-alt-sshd-setup.sh
:: Synchronising package databases...
 mingw32 is up to date
 mingw64 is up to date
 msys is up to date
warning: openssh-8.3p1-1 is up to date -- skipping
warning: cygrunsrv-1.62-2 is up to date -- skipping
warning: mingw-w64-x86_64-editrights-1.03-3 is up to date -- skipping
 there is nothing to do
The command completed successfully.

passwd: unknown user sshd_server_msys2
```

I also tried on my personal computer and it works as a charm, as expected. ¯\_(ツ)_/¯

**Comment by rkitover** [comment]
@mundodisco8 we are kind of debugging blindly, so this is to be expected.

I may have fixed that problem, or not, and there may be other problems.

So please try again, and it may or may not work.

But I have an idea.

I'm going to install Windows Server Essentials on an isolated network
with my Windows 10 VM and set it up as a domain controller.

I need to do this anyway, because I use a windows server for another
project and I need to replace it with this version of windows server,
so doing a trial run of an install will help.

**Comment by mundodisco8** [comment]
That did the trick! Thanks for your support!  

I have now an ssh problem that I can't troubleshoot. I'm feeling like I'm abusing your generosity, but having reached so far... 
This is what I did:
1. Created an ssh key in MSYS following your instructions
1. Added it to authorized_keys in MSYS. Double checked the contents, and it contains the key stored in id_rsa.pub, as expected.
1. Copy those keys to my windows ssh folder, as MSYS and MSYS.pub
1. Added msys to my windows ssh config:
```
Host msys2
  HostName localhost
  Port 2222
  RequestTTY yes
  RemoteCommand MSYSTEM=MSYS exec bash -l
  IdentityFile ~/.ssh/MSYS
```

5. Changed the terminal command to call MSYS to add verbose output. This is the output. It doesn't explicitly say that it refuses my key, but it asks for a password to log in

```
OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
debug1: Reading configuration data C:\\Users\\MYUSER/.ssh/config
debug1: C:\\Users\\MYUSER/.ssh/config line 1: Applying options for msys2
debug1: Connecting to localhost [::1] port 2222.
debug1: Connection established.
debug1: identity file C:\\Users\\MYUSER/.ssh/MSYS type 0
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\MYUSER/.ssh/MSYS-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_for_Windows_7.7
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.3
debug1: match: OpenSSH_8.3 pat OpenSSH* compat 0x04000000
debug1: Authenticating to localhost:2222 as 'MYDOMAIN\\MYUSERNAMEINTHATDOMAIN'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ecdsa-sha2-nistp256 SHA256:zgzfd1IuoQkPsAMh0T4yoHwpqClyZ+3R34CPoT/Wrq0
debug1: Host '[localhost]:2222' is known and matches the ECDSA host key.
debug1: Found key in C:\\Users\\MYUSER/.ssh/known_hosts:8
debug1: rekey after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey after 134217728 blocks
debug1: pubkey_prepare: ssh_get_authentication_socket: No such file or directory
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,sk-ssh-ed25519@openssh.com,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey,password,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Offering public key: RSA SHA256:SOMESHASTUFFIMHIDDINGJUSTINCASE C:\\Users\\MYUSER/.ssh/MSYS
debug1: Authentications that can continue: publickey,password,keyboard-interactive
debug1: Next authentication method: keyboard-interactive
debug1: Authentications that can continue: publickey,password,keyboard-interactive
debug1: Next authentication method: password
debug1: read_passphrase: can't open /dev/tty: No such file or directory
MYDOMAIN\\MYUSERNAMEINTHATDOMAIN@localhost's password:
debug1: Authentications that can continue: publickey,password,keyboard-interactive
Permission denied, please try again.
debug1: read_passphrase: can't open /dev/tty: No such file or directory
MYDOMAIN\\MYUSERNAMEINTHATDOMAIN@localhost's password:
````

It puzzles me because it seems to pick the correct key, and offer it, but I don't see any rejection messages in the log. If I use a different key on purpose, it happens the same, it doesn't explicitly say "that key is wrong, get out of my lawn!"

**Comment by rkitover** [comment]
> I have now an ssh problem that I can't troubleshoot. I'm feeling like I'm abusing your generosity, but having reached so far...

No I'm happy to help because I want this to work.

This looks like another problem because of the domain.

Some things to try:

First, as the msys2 user, run:

```bash
passwd -R
```

and type in your Windows password. This is suggested for domain users by the author of the original script.

If that doesn't work:

In `/etc/ssh/sshd_config` set:

```
LogLevel DEBUG
```

Restart the service:

```bash
net stop msys2_sshd
net start msys2_sshd
```

Make sure you have the `/var/log/msys2_sshd.log` log file, watch it with `tail -f` while trying to log in.

Let me know what you see in that log.

**Comment by mundodisco8** [comment]
```
Invalid user MYDOMAIN\\MYSUER from ::1 port 50697
Failed none for invalid user MYDOMAIN\\MYSUER from ::1 port 50697 ssh2
Failed password for invalid user MYDOMAIN\\MYSUER from ::1 port 50697 ssh2
```

The first line is when the system tries with the key, and the second is after hitting enter when asked for password (empty pass, so it fails). Tried with my account's password, and didn't work either (third line).

**Comment by rkitover** [comment]
Please run the script again, I did some more research (which I should have done initially) and made some more tweaks.

It may turn out that a domain service account is necessary, I hope not because that may not be possible to do with this script.

**Comment by mundodisco8** [comment]
Same result, I'm afraid 😫

```
Invalid user MYDOMAIN\\MYSUER from ::1 port 56255
```

It seems, as you said, that It won't work on a domain

**Comment by rkitover** [comment]
Try some different values for the `User` field in `~/.ssh/config` on the Windows side:

`MYDOMAIN+myuser`

or just

`myuser`

try both upper and lower case for the user name.

**Comment by mundodisco8** [comment]
Tried the following, setting the variable "User" in the `config` file to:

   - myuser
   - MYUSER

   - mydomain+myuser
   - mydomain+MYUSER
   - MYDOMAIN+myuser
   - MYDOMAIN+MYUSER

   - myuser@mydomain
   - myuser@MYDOMAIN
   - MYUSER@mydomain
   - MYUSER@MYDOMAIN

   - MYDOMAIN\\MYUSER (the \ was escaped, so it came out as
   MYDOMAIN\\\\MYUSER)
   - MYDOMAIN\MYUSER(came out as MYDOMAIN\\MYUSER)
   - MYDOMAIN\myuser

   - I tried issuing the command `ssh myuser@mydomain@msys -v` and   `ssh
   MYUSER@MYDOMAIN@msys -v`
   - I tried specifying the key too, but same thing

I think we can stop wasting your time here. If you ever happen to set that
server up, and find something, just write here and I'll pick the
notification. I know how frustrating must be, but I don't want to waste
your time on this. THANKS FOR THE HELP! You really went above and beyond!

**Comment by rkitover** [comment]
> If you ever happen to set that server up, and find something, just write here and I'll pick the notification.

That's the next step, and I will do this soon, this is actually not too onerous a task with virt-manager, virt-manager is quite awesome.

By the way, can you give me the server log for your attempts with different user names, I want to see if the messages are any different.

**Comment by mundodisco8** [comment]
So! I tried again to capture the different usernames and their log output, tried the first one, went to the password request as usual, and noticed it didn't create new log lines... I thought that was weird, so I tried my password and there you go! it worked.

The username I used was just "myusername", low case, without the "mydomain\" bit.

I thought it was odd that it still didn't let me use my key, so tried using my id_rsa key (i was using one named MSYS2 before) and passwordless login worked! 

So then I checked the keys, for the nth time, and they looked the same. And then I decided to do a full compare with VS Code (maybe 1 byte was different?). Turns out that the first line was the same in both keys, so that's why I thought they were the same, but the rest of they key wasn't! I'm such a dum dum!

I'm tempted to repeat the whole process to confirm it just works with your instructions on the Sandbox environment but the results of `whoami` are different, so I don't know if the whole "domain" thing still applies, but I would say that it now works as promised in the tin. 

Thanks!

![tenor](https://user-images.githubusercontent.com/8402707/86478357-0f46a280-bd42-11ea-81ef-5520dff87427.gif)

**Comment by rkitover** [comment]
Well, I am glad that it works. If you don't mind, I'd like your help in putting some finishing touches on the script so that it is fully automatic for domain users.

We have taken care of all the issues aside from the username mapping, and I really appreciate your patience in debugging this.

I will ask you to run some commands and give me the output, you can change the specific names, I just need the format of the strings.

After which, I will make some final adjustments and ask you to backup your configuration and run the script again, and check that automatic login works as expected.

In powershell or cmd, run:

```powershell
whoami
```

In MSYS2 run these commands and give me the output of each:

```bash
whoami

getent passwd $USER | sed 's/,.*//'

getent -w passwd $USER | sed -E 's/(.*):.*/\1/'

id | awk '{ print $1 " " $2 }'

getent group $USER | cut -f1,3 -d:

getent -w group $USER | cut -f1-3 -d:

mkpasswd -c | cut -f1-4 -d:

mkgroup -c | cut -f1,3 -d:
```

**Comment by mundodisco8** [comment]
I sent you a mail

cheers

**Comment by rkitover** [comment]
Awesome, thank you.

So what I gathered from this is that we just need to set `User <posix-user-name>`, and I made this update.

Also fixed a couple of other minor issues, apparently my last couple of improvements broke Cygwin support.

So if you could do a final test for me, I would really appreciate it.

Just move your windows profile `~/.ssh` and your MSYS2 `~/.ssh` out of the way and run the script again, then check if you can login automatically.

**Comment by rkitover** [comment]
@mundodisco8 Just wanted to mention, that you don't have to redo your configuration to test this, just do this:

```bash
mv .ssh ssh-bak
```

in both your Windows profile and the MSYS2 home directory, then once you've checked that the generated configuration works (or doesn't) just do:

```bash
mv ssh-bak .ssh
```

to restore your own configuration.

If you could do this final test for me I would really appreciate it.

**Comment by mundodisco8** [comment]
Damn! I wrote a very long reply, but it seems that I didn't send it...

Long story short, it worked.

- If I set a password when prompted, the script works as expected. I get an error, though, during the installation

```
# bash msys2-alt-sshd-setup.sh
:: Synchronising package databases...
 mingw32                                     468.0 KiB   567 KiB/s 00:01 [#######################################] 100%
 mingw32.sig                                 119.0   B   116 KiB/s 00:00 [#######################################] 100%
 mingw64                                     470.7 KiB   836 KiB/s 00:01 [#######################################] 100%
 mingw64.sig                                 119.0   B   116 KiB/s 00:00 [#######################################] 100%
 msys is up to date
warning: openssh-8.3p1-1 is up to date -- skipping
warning: cygrunsrv-1.62-2 is up to date -- skipping
warning: mingw-w64-x86_64-editrights-1.03-3 is up to date -- skipping
 there is nothing to do
The command completed successfully.

Updating property(s) of '\\MY_LAPTOP_NAME\ROOT\CIMV2:Win32_UserAccount.Domain="MYLAPTOPNAME",Name="sshd_server_msys2"'
Property(s) update successful.
The command completed successfully.


Deleted 1 rule(s).
Ok.

Ok.

processed file: C:\Users\joel.santos\.ssh\id_rsa
Successfully processed 1 files; Failed processing 0 files
processed file: C:\Users\joel.santos\.ssh\id_rsa
Successfully processed 1 files; Failed processing 0 files
processed file: C:\Users\joel.santos\.ssh\id_rsa
Successfully processed 1 files; Failed processing 0 files
ERROR: The system was unable to find the specified registry key or value.
The MSYS2 sshd service is starting.
The MSYS2 sshd service was started successfully.

Please enter your Windows domain user password, this is needed for passwordless logon with a key.

This functionality stores a password in the registry for usage by services
which need to change the user context and require network access.  Typical
applications are interactive remote logons using sshd, cron task, etc.
This password will always tried first when any privileged application is
about to switch the user context.

Note that storing even obfuscated passwords in the registry is not overly
secure.  Use this feature only if the machine is adequately locked down.
Don't use this feature if you don't need network access within a remote
session.

You can delete the stored password by specifying an empty password.

Enter your current password:
Re-enter your current password:
```

- If I don't set a password, then the script doesn't work. Starting a new tab with MSYS in the New Terminal just closes. I assumed (correct me if I'm wrong) that by not filling the password here, then I would get a password request every time I start a new MSYS tab, but doesn't seem to be the case

```
OpenSSH_for_Windows_7.7p1, LibreSSL 2.6.5
debug1: Reading configuration data C:\\Users\\joel.santos/.ssh/config
debug1: C:\\Users\\joel.santos/.ssh/config line 2: Applying options for msys2
debug1: Connecting to localhost [::1] port 2222.
debug1: Connection established.
debug1: identity file C:\\Users\\joel.santos/.ssh/id_rsa type 0
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_rsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_dsa type -1
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_dsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_ecdsa type -1
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_ecdsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_ed25519 type -1
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_ed25519-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_xmss type -1
debug1: key_load_public: No such file or directory
debug1: identity file C:\\Users\\joel.santos/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_for_Windows_7.7
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.3
debug1: match: OpenSSH_8.3 pat OpenSSH* compat 0x04000000
debug1: Authenticating to localhost:2222 as 'joel.santos'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ecdsa-sha2-nistp256 SHA256:SOME_SHA
debug1: Host '[localhost]:2222' is known and matches the ECDSA host key.
debug1: Found key in C:\\Users\\joel.santos/.ssh/known_hosts:1
debug1: rekey after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: rekey after 134217728 blocks
debug1: pubkey_prepare: ssh_get_authentication_socket: No such file or directory
debug1: SSH2_MSG_EXT_INFO received
debug1: kex_input_ext_info: server-sig-algs=<ssh-ed25519,sk-ssh-ed25519@openssh.com,ssh-rsa,rsa-sha2-256,rsa-sha2-512,ssh-dss,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,sk-ecdsa-sha2-nistp256@openssh.com>
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey,password,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Offering public key: RSA SHA256:A_DIFFERENT_SHA_THAN_THE_PREVIOUS_ONE C:\\Users\\joel.santos/.ssh/id_rsa
Connection closed by ::1 port 2222

[process exited with code 255]
```

The log in  `/var/log/msys2_sshd.log`

```
seteuid 1054344: No such device or address
```

Also, in the instructions, the curl command fails in cmd.exe, as it says that `'https` is not a valid protocol. Removing the single quote, or substituting by double quotes, works in MSYS and cmd.exe. Powershell fails because it doesn't accept the `LO` flag

```
Invoke-WebRequest : A parameter cannot be found that matches parameter name 'LO'.
At line:1 char:6
+ curl -LO 'https://raw.githubusercontent.com/rkitover/windows-alt-sshd ...
+      ~~~
    + CategoryInfo          : InvalidArgument: (:) [Invoke-WebRequest], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Microsoft.PowerShell.Commands.InvokeWebRequestCommand
```

**Comment by rkitover** [comment]
Awesome, I really appreciate your help in getting this working.

I think we can consider this fully resolved now, and I have taken care of the remaining minor issues.

> ERROR: The system was unable to find the specified registry key or value.

Here it's just checking the registry key that indicates that developer mode is enabled, if it is enabled we enable real symlink support.

I have suppressed the error message.

I recommend turning on developer mode by the way for this reason.

> Also, in the instructions, the curl command fails in cmd.exe, as it says that 'https is not a valid protocol. Removing the single quote, or substituting by double quotes, works in MSYS and cmd.exe. Powershell fails because it doesn't accept the LO flag

I changed the quotes to double quotes.

In the case of powershell, I use either `powershell-core` or `powershell-preview` from chocolatey, which is a much newer version than the powershell that comes with windows. But I should really test powershell instructions with windows powershell.

In this case, powershell 7 removes the curl alias in favor of the windows executable, while windows powershell does not.

I have added separate instructions for windows powershell and bash/cmd.exe.

Again, thank you very much for helping me debug this, thanks to your help I even managed to avoid setting up a virtual domain controller.

We can now consider this fully resolved!
